packages/expo-router/src/fork/extractPathFromURL.ts

```diff
@@ -92,10 +92,21 @@ export function extractExpoPathFromURL(url: string = '') {
   return extractExactPathFromURL(url).replace(/^\//, '');
 }
 
+// Preview QR codes from expo.dev take on the form of http://u.expo.dev/<uuid>/group/<uuid>
+const expoPreviewLinkRegExp = new RegExp(
+  /^[0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12}\/group\/[0-9A-F]{8}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{4}-[0-9A-F]{12}/i
+);
+
 export function adjustPathname(url: { hostname?: string | null; pathname: string }) {
   if (url.hostname === 'exp.host' || url.hostname === 'u.expo.dev') {
-    // drop the first two segments from pathname:
-    return url.pathname.split('/').slice(2).join('/');
+    if (expoPreviewLinkRegExp.test(url.pathname)) {
+      // drop the /<uuid>/group/<uuid> segments from pathname:
+      return url.pathname.split('/').slice(3).join('/');
+    } else {
+      // drop the /<uuid>/group/<uuid> segments from pathname:
+      // drop the first two segments from pathname:
+      return url.pathname.split('/').slice(2).join('/');
+    }
   }
```

Comment:
This looks rather brittle to me, evidenced by the fact that this bug occurred. There is high spatial coupling between the website and the router, and ideally the router would be agnostic to EAS.

This is my understanding of how the deep linking URLs work:

- In Expo Go, we use a path segment `/--` to separate the URL of the update to load and the string that we want the developer's application to treat as the deep linking path. An example is clearer: what would be `myapp://hello` in a build of the developer's own app would be `exp://u.expo.dev/{projectID}/--/hello` for Expo Go.
- The [Linking](https://docs.expo.dev/guides/linking/#parsing-urls) module takes care of this.

I'm not sure how Expo Dev Client in a development build handles URLs. My guess is that it's similar to Expo Go, but also supports `myapp://hello` if an update is already running.

So, I'd look at using the `Linking` module to see if it can remove the need to check for EAS-specific URLs altogether.